### PR TITLE
Fix crashes not being reported during garbage collection

### LINKED ISSUE #560: Garbage collection crash not reported
### Environment

How do you use Sentry?
Sentry SaaS (sentry.io)

Which version of the SDK?
[0.17.1](https://github.com/getsentry/sentry-unreal/releases/tag/0.17.1)

How did you install the package? (Git-URL, Assetstore)
Git

Which version of Unreal?
5.3.2

Is this happening in Unreal (editor) or on a player like Android, iOS, Windows?
Windows

### Steps to Reproduce

1. Trigger a crash during garbage collection

### Expected Result

Crashes should be reported.

### Actual Result

No crash is reported. The last line in the log is ``LogSentrySdk: invoking `on_crash` hook``.

What I think is happening here is that it fails on [this line in `SentrySubsystemDesktop::HandleBeforeSend()`](https://github.com/getsentry/sentry-unreal/blob/2fe8b5f8b5cfceac65155e9bcfbbb4d084a1de72/plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp#L62), because it is attempting to create a new UObject, but the engine does not allow you to create UObjects during garbage collection. Which means that any crashes that occur during GC would fail to send.

### Any logs or screenshots
This is the relevant tail of the log output.
```
[2024.05.14-18.40.10:855][412]LogOutputDevice: Warning: 

Script Stack (0 frames) :

[2024.05.14-18.40.10:855][412]LogWindows: Error: appError called: Fatal error: [File:D:\******\UE5\Engine\Source\Runtime\CoreUObject\Private\UObject\GarbageCollectionVerification.cpp] [Line: 155] 
Encountered 8 object(s) breaking Disregard for GC assumptions. Please check log for details.



[2024.05.14-18.40.10:855][412]LogWindows: Windows GetLastError: The operation completed successfully. (0)
[2024.05.14-18.40.10:855][412]LogSentrySdk: flushing session and queue before crashpad handler
[2024.05.14-18.40.10:855][412]LogSentrySdk: Verbose: invoking `on_crash` hook
```

### PR DESCRIPTION
**Pull request by tustanivsky**
Closes #560

### CODE REVIEW

File: plugin-dev/Source/Sentry/Private/Desktop/SentryEventDesktop.cpp [hunk 1]
```diff
@@ -56,7 +57,7 @@ ESentryLevel SentryEventDesktop::GetLevel() const
 
 bool SentryEventDesktop::IsCrash() const
 {
-	return false;
+	return IsCrashEvent;
```
Comment (bitsandfoxes) [inline plugin-dev/Source/Sentry/Private/Desktop/SentryEventDesktop.cpp:60]: That is pretty huge! But I don't get why now? Should that have worked previously? I see it being set in the `HandleBeforeCrash` but it seems unrelated to the GC.

File: plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp [hunk 2]
```diff
@@ -73,7 +80,26 @@ sentry_value_t HandleBeforeCrash(const sentry_ucontext_t *uctx, sentry_value_t e
...
+
+		return SentrySubsystem->GetBeforeSendHandler()->HandleBeforeSend(EventToProcess, nullptr)
+			? event
+			: sentry_value_new_null();
+	}
+	else
+	{
+		// If crash occured during garbage collection we can't just obtain a GC lock like with normal events
+		// since there is no guarantee it will be ever freed. In this case crash event will be reported
+		// without calling a `beforeSend` handler.
```
Comment (bitsandfoxes) [inline plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp:100]: That's a pretty huge gotcha that we should put on the docs.

**Review by bitsandfoxes — COMMENTED** [review]
Looking good!

File: plugin-dev/Source/Sentry/Private/Desktop/SentryEventDesktop.cpp (hunk 1 above)
Comment (tustanivsky) [reply plugin-dev/Source/Sentry/Private/Desktop/SentryEventDesktop.cpp:60]: The fact that `IsCrash()` method for desktop events was simply a stub caught my eye only now while reworking `onCrash` handler. We don't really use its Win/Linux implementation within the SDK so I assume that's the reason why it remained unnoticed for so long.

Would you prefer to fix this in a separate PR or we're okay with merging it with other changes here?

File: plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp (hunk 2 above)
Comment (tustanivsky) [reply plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp:100]: How about adding this to the "Known limitations" section in readme instead?

Comment (bitsandfoxes) [reply plugin-dev/Source/Sentry/Private/Desktop/SentrySubsystemDesktop.cpp:100]: Sounds good to me! Thanks!

**Review by bitsandfoxes — APPROVED** [review]

**Pull request merged**